File: src/bloomberg/as/reverse_linked_list_2.cc
#include <bits/stdc++.h>
using namespace std;
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};
/*
[1,2,3,4,5]
2
4
[5]
1
1
[1,2,3]
1
3
*/
class Solution {
  public:
    ListNode *reverseBetween(ListNode *head, int left, int right) {
        if(left == right)
            return head;
        left--;
        right--;
        ListNode *node = head;
        ListNode *l_edge = nullptr;
        for(int i = 0; i < left; i++) {
            l_edge = node;
            node = node->next;
        }
        ListNode *left_origin = node;
        ListNode *next_node = node->next;
        for(int i = left; i < right; i++) {
            ListNode *next2 = next_node->next;
            next_node->next = node;
            node = next_node;
            next_node = next2;
        }
        if(l_edge != nullptr) {
            l_edge->next = node;
        }
        left_origin->next = next_node;
        if(left > 0)
            return head;
        return node;
    }
};

File: src/serial/reverse_linked_list2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseBetween(self, head: Optional[ListNode], left: int, right: int) -> Optional[ListNode]:
        if left == right:
            return head
        
        sentinel = ListNode(0, head)
        rev_head = head
        prev = sentinel
        for _ in range(left-1):
            prev = rev_head
            rev_head = rev_head.next
        
        tail = rev_head
        for i in range(right - left):
            tail = tail.next
        
        third_head = tail.next
        tail.next = None
        # rev_head -> 2
        # tail -> 4
        # prev -> 1
        # third_head -> 5

        # revese [rev_head,,,tail]
        node = rev_head
        peek = node.next
        node.next = None
        while peek:
            temp = peek.next
            peek.next = node
            node = peek
            peek = temp
        
        prev.next = tail
        rev_head.next = third_head
        return sentinel.next
